doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
+ * Keys can be expired from the bucket based on a TTL
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * read-after-write safety unless read replicas used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports
+ * Key starting with `_kv` is reserved for internal use
+ * CLI tool to manage the system as part of `nats`, compatible with client implementations
+ * Read-Only and Read-Write interfaces
+
+Planned features:
+
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * A read-only client can talk directly, and only, to a regional read replica
+ * Discoverability of read replicas per cluster based on in-bucket configuration values
+ * Standard encoders/decoders to deliver zero-trust end to end encryption, perhaps using nkeys
+ * Pluggable storage backends
+
+## Data Types
+
+### Result
+
+This is the result made available over watchers, `Get()` etc. All backends must return an implementation providing at least this information.
+
+```go
+type Result interface {
+	// Bucket is the bucket the data was loaded from
+	Bucket() string
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() string
+	// Created is the time the data was received in the bucket
+	Created() time.Time
+	// Sequence is a unique sequence for this value
+	Sequence() uint64
+	// Delta is distance from the latest value. If history is enabled this is effectively the index of the historical value, 0 for latest, 1 for most recent etc.
+	Delta() uint64
+	// OriginCluster is the cluster where this data originate from, may be empty if sharing was not enabled
+	OriginCluster() string
+	// Operation is the kind of operation this result represents, enum of PUT and DEL
+	Operation() Operation
+}
+```
+
+### Status
+
+This is the status of the KV as a whole
+
+```go
+type Status interface {
+	// Bucket the name of the bucket
+	Bucket() string
+
+	// Values is how many messages are in the bucket, including historical values
+	Values() uint64
+
+	// History returns the configured history kept per key
+	History() int64
+
+	// TTL is how long the bucket keeps values for
+	TTL() time.Duration
+
+	// Cluster returns the name of the cluster holding the read replica of the data
+	Cluster() string
+
+	// Replicas returns how many times data in the bucket is replicated at storage
+	Replicas() (ok int, failed int)
+
+	// Keys returns a list of all keys in the bucket - not possible now
+	Keys() ([]string, error)
+
+	// BackingStore is a backend specific name for the underlying storage - eg. stream name
+	BackingStore() string
```

Comment:
informational...you migth not, or you might seems harmless to expose

Reply:
Could be the store engine? So in future could say sled or redis etc?

Reply:
sure and sled might have a internal name.

Here for example, I know I am using a jetstream instance and maybe I want to backup the stream.  Hows a user to know, I dont want to tell them and expect them to remember that its `KV_BUCKETNAME`, no, I say is this a JS backend? yes, ok backup the stream from BackingStore().

Reply:
This should also not be a string, but deferring to the constants/types for the current storage `StorageType` - I think we need to align some of these fields to the JetStream types they telegraph, avoiding yet more types.

Reply:
The backingstore here would not be something like file - here it would be the stream name, or maybe the consult backend name or etcd endpoint url or whatever makes sense for the backend to give hints to what it is.

Perhaps this could be something like a url even - `jetstream://KV_BOB`  and `etcd://some.node:123` etc, this way the CLI can figure out how to do bakcups etc?